### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by aedsinger**
This PR 
* Refactors the code into a stretch_factory library so that utility functions can be used across production software (RDK), RE1 tools, etc
* incorporates a number of improvements to to RE1_firmware_updater (see Gist https://gist.github.com/hello-binit/0bec03e541baff259cbabcdf960c5298). 
** RE1_firmwmare_updater.py now manages installation of the correct version of the arduino-cli

### CODE REVIEW

**Comment by aedsinger** [comment]
RE1_firmware_updater.py now uses the ```supported_protocols``` field of Devices, so newer version of Stretch Body is required that handles dynamic protocols.

**Review by hello-binit — CHANGES_REQUESTED** [review]
The reduced output is a lot easier to work with. I wasn't able to get the install_path option to work, and I had a few other small issues.

File: python/stretch_factory/firmware_updater.py [hunk 1]
```diff
@@ -0,0 +1,712 @@
...
+                            vt=vs[id]
+                        else:
+                            click.secho('Invalid ID', fg="red" )
+                    print('Selected version %s for device %s'%(vt,device_name))
+                    self.target[device_name]=vt
+        print('')
+        print('')
+        return self.do_update()
+
+    def do_update_to_path(self,path_name):
```
[the remark this answers is not in the document]
Comment (hello-binit) [reply python/stretch_factory/firmware_updater.py:457]: Seems like this affects `--install_branch` too. When the firmware version matches what is currently on-board, it doesn't perform the flash.

```bash
hello-binit@stretch-re1-1002:~$ RE1_firmware_updater.py --arm --install_branch
Collecting information...
.....

################### Select target branch ###################
Collecting information...
....

0: origin/master
1: origin/feature/validate_segments
2: origin/feature/waypoint_trajectories_py3

Please enter desired branch id [0]: 1
Selected branch origin/feature/validate_segments
########################################## UPDATING FIRMWARE TO... ###########################################
HELLO-MOTOR-ARM           | Already at target of Stepper.v0.0.2p1    
System is up to date. No updates to be done
hello-binit@stretch-re1-1002:~$
```

File: python/setup.py [hunk 2]
```diff
@@ -6,23 +6,27 @@
...
     scripts = stretch_scripts,
-    packages=setuptools.find_packages(),
+    packages=['stretch_factory'],
     classifiers=[
         "Programming Language :: Python :: 2",
         "Operating System :: OS Independent",
         "License :: OSI Approved :: GNU General Public License v3 (GPLv3)"
-    ]
+    ],
+    install_requires=['pyserial','pyusb','gitpython']
```
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/setup.py:28]: Done

File: python/stretch_factory/firmware_updater.py [hunk 3]
```diff
@@ -0,0 +1,712 @@
...
+                    dd=stretch_body.wacc.Wacc()
+                elif device == 'hello-pimu':
+                    dd = stretch_body.pimu.Pimu()
+                else:
+                    dd=stretch_body.stepper.Stepper('/dev/'+device)
+                dd.startup()
+                if dd.board_info['firmware_version'] is not None: #Was able to pull board info from device
+                    self.config_info[device]={}
+                    self.config_info[device]['board_info'] = dd.board_info.copy()
+                    self.config_info[device]['supported_protocols']=dd.supported_protocols.keys()
```
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:128]: Supports the older Stretch Body's now as well

File: python/stretch_factory/firmware_updater.py [hunk 4]
```diff
@@ -0,0 +1,712 @@
...
+            return False
+
+    def wait_on_device(self,device_name,timeout=10.0):
+        #Wait for device to appear on bus for timeout seconds
+        print('Waiting for device %s to return to bus.'%device_name)
+        ts=time.time()
+        itr=0
+        while(time.time()-ts<timeout):
+            if self.is_device_present(device_name):
+                print('\n')
```
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:642]: Removed

File: python/stretch_factory/firmware_updater.py [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+                return True
+            itr=itr+1
+            if itr % 5 == 0:
+                sys.stdout.write('.')
+                sys.stdout.flush()
+            time.sleep(0.1)
+        print('\n')
```
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:649]: Removed

File: python/tools/RE1_firmware_updater.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        u.fw_recommended.pretty_print()
+        print('')
+        print('')
+        u.do_update()
+    elif args.install_version:
+        u.do_update_to()
+    elif args.install_branch:
+        u.do_update_to_branch()
+    elif args.install_path:
+        u.do_update_to_path(args.install_path)
```
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/tools/RE1_firmware_updater.py:128]: Done

File: python/stretch_factory/firmware_updater.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:457]: Fixed relative path handling.

[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:457]: It now 'reinstalls' the firmware if it matches what is currently on-board. For development this is more useful, for users hopefully they'd never do this.

[the remark this answers is not in the document]
Comment (aedsinger) [reply python/stretch_factory/firmware_updater.py:457]: I found that a last minute change I made broke the --install_path code. Working again. Also it now checks if the compile succeeded before proceeding.

**Review by hello-binit — APPROVED** [review]
Ready to merge! Firmware updater tool works reliably well.

**Pull request merged**